src/i_oalsound.c

```diff
@@ -352,9 +352,242 @@ const char **I_OAL_GetResamplerStrings(void)
+    static ALuint uiEffectSlot = AL_INVALID;
+    static ALuint uiEffect = AL_INVALID;
+    static ALuint uiFilter = AL_INVALID;
+
+    if (!alGenAuxiliaryEffectSlots ||
+        !alDeleteAuxiliaryEffectSlots ||
+        !alIsAuxiliaryEffectSlot ||
+        !alGenEffects ||
+        !alDeleteEffects ||
+        !alIsEffect ||
+        !alEffecti ||
+        !alEffectf ||
+        !alAuxiliaryEffectSloti ||
+        !alGenFilters ||
+        !alDeleteFilters ||
+        !alIsFilter ||
+        !alFilteri ||
+        !alFilterf)
+    {
+        return;
+    }
+
+    // Unload all effects first.
+
+    if (alIsAuxiliaryEffectSlot(uiEffectSlot))
+    {
+        for (int i = 0; i < MAX_CHANNELS; i++)
+        {
+            alSourcei(oal->sources[i], AL_DIRECT_FILTER, AL_FILTER_NULL);
+            alSource3i(oal->sources[i], AL_AUXILIARY_SEND_FILTER, AL_EFFECTSLOT_NULL, 0, AL_FILTER_NULL);
+        }
+
+        alDeleteAuxiliaryEffectSlots(1, &uiEffectSlot);
+    }
+
+    if (alIsEffect(uiEffect))
+    {
+        alDeleteEffects(1, &uiEffect);
+    }
+
+    if (alIsFilter(uiFilter))
+    {
+        alDeleteFilters(1, &uiFilter);
+    }
+
+    // Apply equalizer effect.
+
+    if (snd_equalizer != EQ_PRESET_OFF)
+    {
+        if (!alIsAuxiliaryEffectSlot(uiEffectSlot))
+        {
+            alGenAuxiliaryEffectSlots(1, &uiEffectSlot);
+        }
+
+        if (!alIsEffect(uiEffect))
+        {
+            alGenEffects(1, &uiEffect);
+        }
+
+        if (!alIsFilter(uiFilter))
+        {
+            alGenFilters(1, &uiFilter);
+        }
+
+        alEffecti(uiEffect, AL_EFFECT_TYPE, AL_EFFECT_EQUALIZER);
+
+        // AL_LOWPASS_GAIN actually controls overall gain for the filter it's
+        // applied to (OpenAL Effects Extension Guide 1.1, pp. 16-17, 135).
+
+        alFilteri(uiFilter, AL_FILTER_TYPE, AL_FILTER_LOWPASS);
+
+        // Gains vary from 0.126 up to 7.943, which means from -18dB attenuation
+        // up to +18dB amplification, i.e. 20*log10(gain).
+
+        #define DB_TO_GAIN(db) ((ALfloat)BETWEEN(0.126f, 7.943f, pow(10.f, db/20.f)))
+
+        alEffectf(uiEffect, AL_EQUALIZER_LOW_GAIN, DB_TO_GAIN(snd_eq_low_gain));
+        alEffectf(uiEffect, AL_EQUALIZER_MID1_GAIN, DB_TO_GAIN(snd_eq_mid1_gain));
+        alEffectf(uiEffect, AL_EQUALIZER_MID2_GAIN, DB_TO_GAIN(snd_eq_mid2_gain));
+        alEffectf(uiEffect, AL_EQUALIZER_HIGH_GAIN, DB_TO_GAIN(snd_eq_high_gain));
+
+        alEffectf(uiEffect, AL_EQUALIZER_LOW_CUTOFF, (ALfloat)snd_eq_low_cutoff);
+        alEffectf(uiEffect, AL_EQUALIZER_HIGH_CUTOFF, (ALfloat)snd_eq_high_cutoff);
+
+        alAuxiliaryEffectSloti(uiEffectSlot, AL_EFFECTSLOT_EFFECT, uiEffect);
+
+        for (int i = 0; i < MAX_CHANNELS; i++)
+        {
+            // Mute the dry path.
+            alFilterf(uiFilter, AL_LOWPASS_GAIN, 0.0f);
+            alSourcei(oal->sources[i], AL_DIRECT_FILTER, uiFilter);
+
+            // Keep the wet path.
+            alFilterf(uiFilter, AL_LOWPASS_GAIN, 1.0f);
```

Comment:
There should be a `snd_eq_preamp` setting that replaces the `1.0f` to prevent clipping when the other EQ settings end up boosting the overall gain.

Reply:
I'm not sure if this applies since OpenAL Soft uses float point calculations internally and then there's a limiter applied by Windows anyway (or the woof.cfg option).

Reply:
I looked at this again and just as a visualization, here's the "bass" preset entered into a separate equalizer program, showing how it needs a -4 dB preamp gain.

https://github.com/fabiangreffrath/woof/assets/56656010/41123718-a234-41ec-ba63-7b84671ef0bd

Reply:
> There should be a `snd_eq_preamp` setting that replaces the `1.0f` to prevent clipping when the other EQ settings end up boosting the overall gain.

Hm, is this on the dB scale? The docs say this value should be in the `[0.1..1.0]` range.

Reply:
> Hm, is this on the dB scale? The docs say this value should be in the `[0.1..1.0]` range.

Yes, it's using dB scale. A gain of 0.001 brings it down -60 dB which I could measure and it seems fine. I added the setting here: https://github.com/fabiangreffrath/woof/pull/1714/commits/be28ee3d4f828a87d039354420bdb39d903bc5d3